### PR TITLE
Projects collection consolidation

### PR DESCRIPTION
**Pull request by fbourigault**
This contains changes for [!8962](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/8962), [!8877](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/8877) and [!9674](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/9674).

I introduced the `OptionsResolver` to validate and sanitize parameters. If we agree with this, the same pattern could be applied everywhere.

### CODE REVIEW

File: lib/Gitlab/Api/Projects.php [hunk 1]
```diff
@@ -2,77 +2,57 @@
...
-            'order_by' => $order_by,
-            'sort' => $sort
-        ));
+     * @param array $parameters
+     * @return mixed
+     */
+    public function all(array $parameters = [])
+    {
+        $resolver = $this->createOptionsResolver();
+        $normalizer = function ($value) {
```
Comment (dbu) [inline lib/Gitlab/Api/Projects.php]: i would call this more specific, maybe $booleanNormalizer

File: lib/Gitlab/Api/Projects.php [hunk 2]
```diff
@@ -2,77 +2,57 @@
...
-     */
-    public function search($query, $page = 1, $per_page = self::PER_PAGE, $order_by = self::ORDER_BY, $sort = self::SORT)
-    {
-        return $this->get('projects/search/'.$this->encodePath($query), array(
-            'page' => $page,
-            'per_page' => $per_page,
-            'order_by' => $order_by,
-            'sort' => $sort
-        ));
+     * @param array $parameters
```
Comment (dbu) [inline lib/Gitlab/Api/Projects.php]: imho this should explain what parameters are valid and what they mean. something like
```
@param array $parameters
    @var bool  archived Whether to include archived projects in the result (if thats what it means)
    @var string visibility  One of public|internal|private
    ...
    (include or reference the options from the base class too)
```

the old methods had a parameter list so you had an idea of what parameters you have, but still can leave you guessing at the meaning.

File: lib/Gitlab/Api/AbstractApi.php [hunk 3]
```diff
@@ -190,6 +191,29 @@ protected function encodePath($path)
         return str_replace('.', '%2E', $path);
     }
 
+    /**
+     * Create a new OptionsResolver with page, per_page and sort options.
+     *
+     * @return OptionsResolver
+     */
+    protected function createOptionsResolver()
```
Comment (dbu) [inline lib/Gitlab/Api/AbstractApi.php]: why is this its own method? its only used once... or is this a first step and more methods will be refactored to this?

Comment (fbourigault) [inline lib/Gitlab/Api/AbstractApi.php]: This will be used everywhere once we agreed on the current implementation.

**Comment by fbourigault** [comment]
Thanks for review!

**Pull request merged**